syntax = "proto3";

option java_multiple_files = true;
option java_package = "org.hopenly.ilyde.grpc";
option java_outer_classname = "ProjectProto";
option objc_class_prefix = "PS";

package project;

// The project service definition.
service ProjectServices {
    // Retrieve a project from id
    rpc Retrieve (ID) returns (Project) {}
    //  Create a project
    rpc Create (Project) returns (Project) {}
    // Update a project
    rpc Update (Project) returns (Project) {}
    // Delete a project from ID
    rpc Delete (ID) returns (Status) {}
    // Search for projects
    rpc Search (SearchProjectRequest) returns (SearchProjectResponse) {}
    // Retrieve a project's revision from id
    rpc RetrieveRevision (ID) returns (Revision) {}
    //  Create a project's Revision
    rpc CreateRevision (Revision) returns (Revision) {}
    // Search for project's revisions
    rpc SearchRevision (SearchRevisionRequest) returns (SearchRevisionResponse) {}
}

// FileVersion
message FileVersion {
    string name = 1;
    string version = 2;
}

// ID message
message ID {
    string id = 1;
}

// message IDE
enum VISIBILITY {
    PRIVATE = 0;
    PUBLIC = 1;
}

// message STATE
enum STATE {
    OPEN = 0;
    CLOSED = 1;
}

// message TEMPLATE
enum TEMPLATE {
    GENERIC = 0;
}

// message Project
message Project {
    string id = 1;
    string name = 2;
    string description = 3;
    VISIBILITY visibility = 4;
    TEMPLATE template = 5;
    string owner = 6;
    repeated string members = 7;
    STATE state = 8;
    string repo_bucket = 9;
    string create_at = 10;
    string last_update = 11;
}

// message Revision
message Revision {
    string id = 1;
    string commit = 2;
    string author = 3;
    string project = 4;
    repeated FileVersion file_tree = 5;
    string create_at = 6;
}

// message Status
message Status {
    uint32 status = 1;
    string message = 2;
}

// message search project Request
message SearchProjectRequest {
    ProjectFilter query = 1;
    uint32 page = 2;
    uint32 limit = 3;
    message  ProjectFilter {
        string id = 1;
        string name = 2;
        VISIBILITY visibility = 3;
        TEMPLATE template = 4;
        string member = 5;
        STATE state = 6;
    }
}

// message search revision Request
message SearchRevisionRequest {
    RevisionFilter query = 1;
    uint32 page = 2;
    uint32 limit = 3;
    message  RevisionFilter {
        string id = 1;
        string project = 2;
        string author = 3;
    }
}

// message  searchResponse
message SearchProjectResponse {
    uint32 total = 1; // search total count results
    uint32 page = 2;
    uint32 limit = 3;
    repeated Project data = 4;
}

// message  searchRevisionResponse
message SearchRevisionResponse {
    uint32 total = 1; // search total count results
    uint32 page = 2;
    uint32 limit = 3;
    repeated Revision data = 4;
}